src/chainstate/burn/distribution.rs

```diff
@@ -38,195 +39,301 @@ use util::log;
+                        key_vtxindex: referenced_commit.key_vtxindex,
+                        key_block_ptr: referenced_commit.key_block_ptr,
+                        block_hash: Hash160::from_sha256(&referenced_commit.block_header_hash.0),
+                    };
+
+                    if let Some(user_burn_recipients) =
+                        user_burn_targets.get_mut(&user_burn_target_key)
+                    {
+                        user_burn_recipients.push(commitment_ix);
+                    } else {
+                        user_burn_targets.insert(user_burn_target_key, vec![commitment_ix]);
+                    }
+
+                    // found a chained utxo, connect
+                    linked_commit[(window_size - 1 - rel_block_height) as usize] =
+                        Some(LinkedCommitmentScore {
+                            op: referenced_commit,
+                            rel_block_height,
+                            user_burns: 0,
+                        });
                 }
             }
         }
 
-        if user_burns.len() > 1 {
-            let block_height = user_burns[0].block_height;
-            for i in 0..user_burns.len() {
-                if user_burns[i].block_height != block_height {
-                    panic!(
-                        "FATAL ERROR: user burn {} is at ({},{}) not {}",
-                        &user_burns[i].txid,
-                        user_burns[i].block_height,
-                        user_burns[i].vtxindex,
-                        block_height
-                    );
+        // next, we need to associate user burns with the leader block commits.
+        //   this is where things start to go a little wild:
+        //
+        //  User burns identify a block commit using VRF public key, so we'll use the
+        //    user_burn_targets map to figure out which linked commitment should receive
+        //    the user burn
+        let mut commit_txid_to_user_burns: HashMap<_, Vec<UserBurnSupportOp>> = HashMap::new();
+
+        // iterate across user burns in block_height order
+        for (rel_block_height, user_burns_at_height) in user_burns.into_iter().enumerate() {
+            for mut user_burn in user_burns_at_height.into_iter() {
+                let UserBurnSupportOp {
+                    key_vtxindex,
+                    key_block_ptr,
+                    block_header_hash_160,
+                    burn_fee,
+                    ..
+                } = user_burn.clone();
+
+                let user_burn_target_key = UserBurnIdentifier {
+                    rel_block_height: rel_block_height as u8,
+                    key_vtxindex: key_vtxindex,
+                    key_block_ptr: key_block_ptr,
+                    block_hash: block_header_hash_160,
+                };
+
+                if let Some(user_burn_recipients) = user_burn_targets.get(&user_burn_target_key) {
+                    let per_recipient = burn_fee / (user_burn_recipients.len() as u64);
```

Comment:
Wait, didn't we already do this in the sortition DB?

Reply:
No -- the function in the sortition db is only ever invoked by the block preprocessing logic. The primary difference between what's happening in that code and what's happening here is that in the other function, the caller wants specifically only the user burns that supported a particular known-to-have-won block. That data is used in matured rewards calculations.

Here, we need to evaluate all the possible user burns over the window.

Reply:
ah, gotcha